SmartPlaces.Facilities/lib/IngestionManager/src/AzureDigitalTwins/AzureDigitalTwinsGraphManager.cs

```diff
@@ -166,43 +167,110 @@ protected virtual async Task ImportRelationshipsAsync(IDictionary<string, BasicR
+                            }
                         }
                         else
                         {
-                            Logger.LogError(ex, "Failed to add relationship: {relationships}", JsonSerializer.Serialize(relationship));
-                            TelemetryClient.GetMetric(failedMetricIdentifier).TrackValue(1, relationship.Name);
+                            try
+                            {
+                                var jsonPatchDocumentHasUpdates = TwinMergeHelper.TryCreatePatchDocument(existingRelationship, relationship, out var jsonPatchDocument);
+
+                                if (jsonPatchDocumentHasUpdates)
+                                {
+                                    await digitalTwinsClient.UpdateRelationshipAsync(relationship.SourceId, relationship.Id, jsonPatchDocument, cancellationToken: cancellationToken);
+                                    TelemetryClient.GetMetric(updateRelationshipSucceededMetricIdentifier).TrackValue(1, relationship.Name);
+                                    relationshipUpdateSuccesses++;
+                                }
+                                else
+                                {
+                                    relationshipUpdateSkips++;
+                                    TelemetryClient.GetMetric(updateRelationshipSkippedMetricIdentifier).TrackValue(1, relationship.Name);
+                                }
+                            }
+                            catch (RequestFailedException ex)
+                            {
+                                relationshipUpdateFailures++;
+
+                                // If we are over the limit, sleep for a bit then add the relationship to the retry list
+                                // Todo: jobee - are there any other cases that need to be retried?
+                                if (ex.Status == 429)
+                                {
+                                    TelemetryClient.GetMetric(updateRelationshipThrottledMetricIdentifier).TrackValue(1);
+                                    Logger.LogError("Connection to ADT Throttled while updating relationship. Sleeping.");
+                                    await Task.Delay(options.RetryDelayInMs, cancellationToken);
+                                    failedRelationships.TryAdd(relationship.Id, relationship);
+                                }
+                                else
+                                {
+                                    Logger.LogError(ex, "Failed to update relationship: {relationship}", JsonSerializer.Serialize(relationship));
+                                    TelemetryClient.GetMetric(updateRelationshipFailedMetricIdentifier).TrackValue(1, relationship.Name);
+                                }
+                            }
                         }
                     }
+                    catch (RequestFailedException ex) when (ex.ErrorCode == "DigitalTwinNotFound")
```

Comment:
This is probably more of a question for the input to this method, is there any filtering that can be done to say get me all relationships where target is not unclassified/point?

Reply:
We're only getting the relationships for classified points already... if the twin doesn't exist, it's more likely that an error has occurred in the mapping and the types we are mapping don't support the relationship specified, or the twin failed to be created during the twin load process and now the relationship is invalid.